existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,194 @@
+# Existential vs universal quantification
+
+## Introduction
+This article gives brief information about one of the most common extension in Haskell projects - `ExplicitForAll`
+and syntax constructions that this extension brings to language.
+Stay tuned if you want to know how to create lists of differently typed variables, how to add complex constraints and more!
+
+From type theory we know about the existence of two quantifiers - `$\forall$` and `$\exists$`.
+The former means: `"For every term x the statement is true"`, the later: `"There is at least one term x for which this statement is true"`.
+
+Since Haskell is based on type theory there is a place for these quantifiers in language syntax, however, in Haskell we are talking not about some terms and statements, but about types.
+By default, all type variables are bound by a `forall` quantifier, which means that every type, that passes constraints, can be used inside this function/type declaration.
+
+```haskell
+id :: forall a. a -> a
+id x = x
+```
+
+You can't see this quantification without using a specific extension - "Explicit Forall".
+
+## Universal quantification and ExplicitForall
+By default, `forall.` is implicit and sometimes it can lead to unpleasant situations.
+Consider this code:
+
+```Haskell
+prepend2 :: a -> [a] -> [a]
+prepend2 x xs = pairFun x ++ xs
+where pairFun y = [y, y]
+```
+This code compiles without any errors, until you try to use variables from the outer function inside the `pairFun` function.
+```Haskell
+-- code compiles with error
+prepend2 :: a -> [a] -> [a]
+prepend2 x xs = pair ++ xs
+where pair :: [a]
+      pair = [x, x]
+```
+
+If you try to compile this one you will get an error because compiler can't match type variable `a` from outer
+declaration, with the one in inner declaration.
+This is happening, because the real signature of this function is this:
+
+```Haskell
+prepend2 :: forall a. a -> [a] -> [a]
+prepend2 x xs = pair ++ xs
+where pair :: forall a. [a]
+      pair = [x, x]
+```
+
+In order to correct the code you need to use `-XScopedTypeVariables` extension, which will only work if you
+explicitly add forall keyword to `prepend2` function.
+
+### Usage
+Let's try to add `forall.` manually.
+The syntax is very simple - at the beginning of function or instance declaration, before any constraints or arguments were introduced use
+quantifier with all type variables you want to introduce later (Note: you get a compilation error if you use undeclared
+type variables after forall):
+
+```Haskell
+func :: forall a b c m. a -> b -> m c -- OK, it compiles
+func :: forall a b c.   a -> b -> m c -- Error, type variable `m` is not introduced
+```
+Here we introduce four variables. We might want to introduce some constraints.
+For example, it is very likely from the function definition that `m` is an instance of `Monad` typeclass.
```

Comment:
> it is very likely from the function definition that `m` is an instance of `Monad` typeclass.

But why a `Monad` ?  Doesn't the signature only tell us that `m` is a type that takes another type as a paramter/argument?

Reply:
Well, when I see an type variable named m as the wrapper over the result, I subconsciously think of it as of some monad. But you are right, I think I need to correct it